core/Piranha.Manager.Core/Services/MediaService.cs

```diff
@@ -119,12 +108,19 @@ public async Task<MediaListModel> GetList(Guid? folderId = null, MediaType? filt
                 }).ToList();
 
             foreach (var folder in model.Folders)
-            {
-                //
-                // TODO: Optimize, we don't need all this data
-                //
-                folder.ItemCount = (await _api.Media.GetAllAsync(folder.Id)).Count() + structure.GetPartial(folder.Id).Count();
-            }
+                folder.ItemCount = await _api.Media.GetCountAsync(folder.Id) +
+                                   structure.GetPartial(folder.Id).Count;
+            if (width.HasValue)
+                foreach (var mp in pairMedia.Where(m => m.media.Type == MediaType.Image))
+                {
+                    mp.mediaItem.AltVersionUrl =
+                        (await _api.Media.EnsureVersionAsync(mp.media, width.Value, height).ConfigureAwait(false))
+                        .TrimStart('~');
```

Comment:
How can this be faster than the original code? The original version returns the list and then lets the browser request the scaled version for the media file meaning that the list view will return fast and the images will be scaled on demand. With this code, first time you hit a folder with a large number of images that hasn't been scaled, getting the list will have to scale all of the images in the folder.

Reply:
I totally get that it makes fewer requests which is faster, but how did you experience performance when entering a folder with **many** images for the first time with no scaled versions? Depending on storage my feeling is this could take a long time.

Reply:
This is a good point, and the first time performance would suffer, although it still only took 1 minute 14 seconds to process 1,777 images on my machine (granted I'm running 8 cores at 4GHz each which many could hosting systems would not have available). my main thought was reducing SQL query time.

Perhaps an alternative would be a hybrid approach. Say that instead it just checks the versions returned with the Media object and sees if any of those match the requested height and width. If it does, return those. Otherwise, it simply returns null for that particular alternate URL. Then the client can load the folder up, and it renders what URLs it already has. What it doesn't it will proceed to request from the server. I would assume that a javascript function could handle this locally after the view loads. We pass it the item object and it checks to see if the thumbnail has returned. If not, then it calls getThumbnailUrl locally and runs like normal.

This way we load the view in a couple seconds either way.

Reply:
I think that’s a great idea, after first load all scaled versions would then be returned directly to the client!